### PR TITLE
FossId: Store the scan code of a scan in the ScanSummary

### PR DESCRIPTION
**Pull request by bs-ondem**
Check commit messages for details.

### CODE REVIEW

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 1]
```diff
@@ -110,6 +110,9 @@ class FossId internal constructor(
         @JvmStatic
         private val WAIT_DELAY = 10.seconds
 
+        @JvmStatic
+        private val SCAN_CODE_KEY = "scancode"
```
Comment (nnobelis) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:114]: Rather plural `scancodes` ?

File: model/src/main/kotlin/ScanSummary.kt [hunk 2]
```diff
@@ -72,7 +72,14 @@ data class ScanSummary(
...
     @JsonInclude(JsonInclude.Include.NON_EMPTY)
-    val issues: List<OrtIssue> = emptyList()
+    val issues: List<OrtIssue> = emptyList(),
+
+    /**
+     * A map for scanner specific data that cannot be mapped into any property of a scan summary, but still needs to be
+     * stored in it.
+     */
+    @JsonInclude(JsonInclude.Include.NON_EMPTY)
+    val additionalData: Map<String, String> = sortedMapOf()
```
[the remark this answers is not in the document]
Comment (nnobelis) [reply model/src/main/kotlin/ScanSummary.kt:82]: @bs-ondem If we use the `ScanContext`, then we have to store the scan codes as CSV like I did with the labels, right ?

Why do you think it isn't a good fit ?

[the remark this answers is not in the document]
Comment (bs-ondem) [reply model/src/main/kotlin/ScanSummary.kt:82]: > @bs-ondem If we use the `ScanContext`, then we have to store the scan codes as CSV like I did with the labels, right ?

Yes, if we use the `labels` in the `ScanContext` then we have to store the scan codes CSV like or in a comma separated list of scan codes.

> Why do you think it isn't a good fit ?

`ScanContext` is only used in relation with the experimental scanners and can't be used currently with the `--scanner` option of the ORT scan command. In addition, a scan code belongs to a specific scan of package which should be mapped in the `ScanSummary` of the package.

File: model/src/main/kotlin/ScanResult.kt [hunk 3]
```diff
@@ -41,7 +42,14 @@ data class ScanResult(
     /**
      * A summary of the scan results.
      */
-    val summary: ScanSummary
+    val summary: ScanSummary,
+
+    /**
+     * A map for scanner specific data that cannot be mapped into any generalized property, but still needs to be
+     * stored for in the scan result.
```
Comment (nnobelis) [inline model/src/main/kotlin/ScanResult.kt:49]: - for

File: model/src/main/kotlin/ScanResult.kt [hunk 4]
```diff
@@ -20,6 +20,7 @@
 package org.ossreviewtoolkit.model
 
 import com.fasterxml.jackson.annotation.JsonIgnoreProperties
+import com.fasterxml.jackson.annotation.JsonInclude
```
Comment (sschuberth) [inline model/src/main/kotlin/ScanResult.kt:23]: Commit message nits:
- check line wrapping
- quote `additionalData`
- don't only say it's called `additionalData` now, but also why (to emphasize this is about optional data that cannot be captures in the common data model).
- let's say "*will be used" instead of "can be used", so make clear that there is a real need for this *now*

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 5]
```diff
@@ -337,10 +340,20 @@ class FossId internal constructor(
                             severity = Severity.HINT
                         )
                         val summary = ScanSummary(
-                            startTime, Instant.now(), "", sortedSetOf(), sortedSetOf(), listOf(issue)
+                            startTime,
+                            Instant.now(),
+                            "",
+                            sortedSetOf(),
+                            sortedSetOf(),
+                            listOf(issue)
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:348]: Nit: Unrelated / unnecessary change.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 6]
```diff
@@ -337,10 +340,20 @@ class FossId internal constructor(
...
                         )
 
-                        val scanResult = ScanResult(provenance, details, summary)
+                        val scanResult = ScanResult(
+                            provenance,
+                            details,
+                            summary,
+                            mapOf(SCAN_CODE_KEY to scanCode)
+                        )
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:356]: Nit: This would still fit into a single line, and probably still is readable enough that way.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt (hunk 1 above)
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:114]: Consider making this `internal` so it can be used in the test below.

File: scanner/src/test/kotlin/scanners/fossid/FossIdTest.kt [hunk 7]
```diff
@@ -686,6 +688,33 @@ class FossIdTest : WordSpec({
...
+                .expectCheckScanStatus(scanCode, ScanStatus.FINISHED)
+                .expectCreateScan(projectCode, scanCode, vcsInfo)
+                .expectDownload(scanCode)
+                .mockFiles(scanCode, identifiedRange = 1..2, markedRange = 1..2)
+
+            val fossId = createFossId(config)
+
+            val results = fossId.scan(listOf(createPackage(pkgId, vcsInfo))).scanResults(pkgId)
+
+            val expectedAdditionalData = mapOf("scancode" to scanCode)
```
Comment (sschuberth) [inline scanner/src/test/kotlin/scanners/fossid/FossIdTest.kt:712]: Use the constant for `"scancode"`, see my comment above.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: model/src/main/kotlin/ScanResult.kt (hunk 4 above)
Comment (sschuberth) [reply model/src/main/kotlin/ScanResult.kt:23]: > quote `additionalData`

My fault, with "quote" I meant "add backticks" to get formatting exactly as in my example; but let's keep it as-is now.

**Review by sschuberth — APPROVED** [review]
